repl/src/main/scala/com/cloudera/livy/repl/Session.scala

```diff
@@ -82,21 +84,46 @@ class Session(interpreter: Interpreter, stateChangedCallback: SessionState => Un
   def execute(code: String): Int = {
     val statementId = newStatementId.getAndIncrement()
     _statements(statementId) = new Statement(statementId, StatementState.Waiting, null)
+
     Future {
-      _statements(statementId) = new Statement(statementId, StatementState.Running, null)
+      _statements(statementId).state.compareAndSet(StatementState.Waiting, StatementState.Running)
 
-      _statements(statementId) =
-        new Statement(statementId, StatementState.Available, executeCode(statementId, code))
+      setJobGroup(statementId)
+      val executeResult = if (_statements(statementId).state.get() != StatementState.Cancelling) {
```

Comment:
Another problem is, if user code is catching SparkException between calls, 2nd and further calls to Spark are not cancelled.
```
try {
  sc... # Call 1
} catch {
   case _ =>
}
try {
  sc... # Call 2
} catch {
   case _ =>
}
```
I think it's fine to not fix this but it should be documented clearly because it's hard to spot.